### PR TITLE
#238: PFADD/PFCOUNT implementation

### PR DESCRIPTION
**Pull request by lucifercr07**
- Adding support for [PFADD](https://redis.io/docs/latest/commands/pfadd/)/[PFCOUNT](https://redis.io/docs/latest/commands/pfcount/) command implementation.
- Hyperloglog implementation library used: [axiomhq/hyperloglog](https://github.com/axiomhq/hyperloglog)
    - Other libraries were analysed like [clarkduvall/hyperloglog](https://github.com/clarkduvall/hyperloglog) and [datadog/hyperloglog](https://github.com/DataDog/hyperloglog) but these are very old and don't seem to be maintained.
- Current implementation only supports `Sparse` representation and doesn't switch to `Dense` once threshold is reached as done by Redis. Also it doesn't do caching as part of `PFCOUNT`.
- **TODO** In follow up PR when implementing `PFMERGE, PFDEBUG` etc.:
    - Look into caching implementation if required for `PFCOUNT`.
    - Benchmark tests for current implementation.
    - Benchmark testing for `Sparse` and `Dense` representations.
    - Analyse feasibility of switch between `Sparse` and `Dense` representations based on benchmark testing.

### CODE REVIEW

File: core/query_watcher.go
Comment (lucifercr07) [inline core/query_watcher.go:1]: Linter changes.

File: core/executor.go
Comment (lucifercr07) [inline core/executor.go:1]: Linter changes.

**Comment by lucifercr07** [comment]
@JyotinderSingh / @soumya-codes please review once. Few considerations have listed as part of description [here](https://github.com/DiceDB/dice/pull/422#issue-2494111108).

**Comment by lucifercr07** [comment]
@JyotinderSingh conflicts resolved. Please review once.

File: tests/hyperloglog_test.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+			commands: []string{"PFADD hll1 foo bar zap", "PFADD hll1 zap zap zap", "PFADD hll1 foo bar", "PFCOUNT hll1"},
+			expected: []interface{}{int64(1), int64(0), int64(0), int64(3)},
+		},
+		{
+			name: "PFADD with multiple keys",
+			commands: []string{
+				"PFADD hll2 foo bar zap", "PFADD hll2 zap zap zap", "PFCOUNT hll2",
+				"PFADD some-other-hll 1 2 3", "PFCOUNT hll2 some-other-hll"},
+			expected: []interface{}{int64(1), int64(0), int64(3), int64(1), int64(6)},
+		},
```
Comment (JyotinderSingh) [inline tests/hyperloglog_test.go:40]: Could you add another test case that verifies behaviour when you call PFCOUNT with multiple HLLs and one or more HLLs don't actually exist.

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Thanks for adding these changes @lucifercr07! The changes look good and well tested, just left one minor command.

**Comment by JyotinderSingh** [comment]
> **TODO** In follow up PR when implementing `PFMERGE, PFDEBUG` etc.:
> * Look into caching implementation if required for `PFCOUNT`.
> * Benchmark tests for current implementation.
> * Benchmark testing for `Sparse` and `Dense` representations.
> * Analyse feasibility of switch between `Sparse` and `Dense` representations based on benchmark testing.

Could you please create Issues for these TODOs to ensure we track them?

**Comment by lucifercr07** [comment]
> > **TODO** In follow up PR when implementing `PFMERGE, PFDEBUG` etc.:
> > 
> > * Look into caching implementation if required for `PFCOUNT`.
> > * Benchmark tests for current implementation.
> > * Benchmark testing for `Sparse` and `Dense` representations.
> > * Analyse feasibility of switch between `Sparse` and `Dense` representations based on benchmark testing.
> 
> Could you please create Issues for these TODOs to ensure we track them?

Sure, will create. Will try to look into few of them as as part of `PFMERGE, PFDEBUG` implementations.

File: tests/hyperloglog_test.go (hunk 1 above)
Comment (lucifercr07) [reply tests/hyperloglog_test.go:40]: Added test case for this scenario.

**Review by JyotinderSingh — APPROVED** [review]
Thanks for adding HLL support to Dice @lucifercr07! The changes look good and well tested.
LGTM.

**Pull request merged**